<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction • disclapmix</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">disclapmix</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">1.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction.html">Introduction</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Introduction</h1>
                        <h4 class="author">Mikkel Meyer Andersen</h4>
            
            <h4 class="date">21 April, 2018</h4>
      
      
      <div class="hidden name"><code>introduction.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>This tutorial introduces the discrete Laplace method implemented in this R package (<code>disclapmix</code>) for estimating Y-STR haplotype frequencies as described by <span class="citation">(Andersen, Eriksen, and Morling 2013)</span>.</p>
<p>First, we demonstrate how to use the open source software for basic tasks as e.g. haplotype frequency estimation and also more advanced tasks as e.g. deconvolution of two person mixtures. First, some packages are loaded <span class="citation">(Andersen and Eriksen 2014)</span>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(disclap)
<span class="kw">library</span>(disclapmix)</code></pre></div>
<p>If you do not have installed the <code>disclap</code> package, please visit <a href="http://cran.r-project.org/package=disclap" class="uri">http://cran.r-project.org/package=disclap</a>.</p>
<p>For reproducibility, the seed for the (pseudo) random number generator is set:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1</span>)</code></pre></div>
</div>
<div id="sec:disclap" class="section level1">
<h1 class="hasAnchor">
<a href="#sec:disclap" class="anchor"></a>The discrete Laplace distribution</h1>
<p>The discrete Laplace distribution is a probability distribution like e.g. the binomial distribution or the normal/Gaussian distribution.</p>
<p>The discrete Laplace distribution has two parameters: a dispersion parameter <span class="math inline">\(0 &lt; p &lt; 1\)</span> and a location parameter <span class="math inline">\(y \in \mathbb{Z} = \{..., -2, -1, 0, 1, 2, ... \}\)</span>.</p>
Let <span class="math inline">\(X \sim DL(p, y)\)</span> denote that the random variable <span class="math inline">\(X\)</span> follows a discrete Laplace distribution with dispersion parameter <span class="math inline">\(0 &lt; p &lt; 1\)</span> and location parameter <span class="math inline">\(y\)</span>. Then a realisation of the random variable, <span class="math inline">\(X = x\)</span>, can be any integer in <span class="math inline">\(\mathbb{Z}\)</span>. The random variable <span class="math inline">\(X\)</span> has the probability mass function given by
<span class="math display">\[\begin{align}
  f(X = x; p, y) = \frac{1-p}{1+p} \cdot p^{\vert x - y \vert} \quad \text{for $x \in \mathbb{Z}$}.
\end{align}\]</span>
<p>As seen, only the absolute value of <span class="math inline">\(x - y\)</span> is used. This means that the probability mass function is symmetric around <span class="math inline">\(y\)</span>.</p>
<p>Let us try to plot the probability mass function <span class="math inline">\(f(X = x; p, y)\)</span> for <span class="math inline">\(p=0.3\)</span> and <span class="math inline">\(y=13\)</span> from <span class="math inline">\(x=8\)</span> to <span class="math inline">\(x=18\)</span>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p &lt;-<span class="st"> </span><span class="fl">0.3</span>
y &lt;-<span class="st"> </span>13L <span class="co"># L for integer type</span>
x &lt;-<span class="st"> </span><span class="kw">seq</span>(8L, 18L, <span class="dt">by =</span> 1L)
<span class="kw">barplot</span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/disclap/topics/ddisclap">ddisclap</a></span>(x -<span class="st"> </span>y, p), <span class="dt">names =</span> x, <span class="dt">xlab =</span> <span class="st">"x, e.g. Y-STR allele"</span>, 
  <span class="dt">ylab =</span> <span class="kw">paste</span>(<span class="st">"Probability mass, f(X = x; "</span>, p, <span class="st">", "</span>, y, <span class="st">")"</span>, <span class="dt">sep =</span> <span class="st">""</span>))</code></pre></div>
<div class="figure">
<img src="introduction_files/figure-html/unnamed-chunk-4-1.png" alt="The probability mass function, $f(X = x; p, y)$, for the discrete Laplace distribution with dispersion parameter $p=0.3$ and location parameter $y=13$ from $x=8$ to $x=18$." width="672"><p class="caption">
The probability mass function, <span class="math inline">\(f(X = x; p, y)\)</span>, for the discrete Laplace distribution with dispersion parameter <span class="math inline">\(p=0.3\)</span> and location parameter <span class="math inline">\(y=13\)</span> from <span class="math inline">\(x=8\)</span> to <span class="math inline">\(x=18\)</span>.
</p>
</div>
<p>We plot the distribution for values of <span class="math inline">\(x\)</span> from <span class="math inline">\(8\)</span> to <span class="math inline">\(18\)</span> as there is almost no probability mass outside these values. We can find out how much of the probability mass that we have plotted:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sum</span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/disclap/topics/ddisclap">ddisclap</a></span>(x -<span class="st"> </span>y, p))</code></pre></div>
<pre><code>## [1] 0.9988785</code></pre>
<p>Thus, only 0.0011215 of the probability mass is outside <span class="math inline">\(\{8, 9, ..., 17, 18\}\)</span>.</p>
If we have a sample of realisations from <span class="math inline">\(X \sim DL(p, y)\)</span> denoted by <span class="math inline">\(\{ x_i \}_{i=1}^n\)</span>, then maximum likelihood estimates are given by the following quantities <span class="citation">(Andersen, Eriksen, and Morling 2013)</span>:
<span class="math display">\[\begin{align}
  \hat{y} &amp;= \text{median} \{ x_i \}_{i=1}^n , \\
  \hat{\mu} &amp;= \frac{1}{n} \sum_{i=1}^n \vert x_i - \hat{y} \vert \, \text{and} \\
  \hat{p} &amp;= \hat{\mu}^{-1} \left (\sqrt{\hat{\mu}^2 + 1} - 1 \right ) .
\end{align}\]</span>
<p>Example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1</span>) <span class="co"># Makes it possible to reproduce the simulation results</span>
p &lt;-<span class="st"> </span><span class="fl">0.3</span> <span class="co"># Dispersion parameter </span>
y &lt;-<span class="st"> </span><span class="dv">13</span> <span class="co"># Location parameter </span>
x &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/disclap/topics/ddisclap">rdisclap</a></span>(<span class="dv">100</span>, p) +<span class="st"> </span>y <span class="co"># Generate a sample using the rdisclap function</span>

y_hat &lt;-<span class="st"> </span><span class="kw">median</span>(x)
y_hat</code></pre></div>
<pre><code>## [1] 13</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mu_hat &lt;-<span class="st"> </span><span class="kw">mean</span>(<span class="kw">abs</span>(x -<span class="st"> </span>y_hat))
mu_hat</code></pre></div>
<pre><code>## [1] 0.57</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p_hat &lt;-<span class="st"> </span>mu_hat^(-<span class="dv">1</span>) *<span class="st"> </span>(<span class="kw">sqrt</span>(mu_hat^<span class="dv">2</span> +<span class="st"> </span><span class="dv">1</span>) -<span class="st"> </span><span class="dv">1</span>)
p_hat <span class="co"># We expect 0.3</span></code></pre></div>
<pre><code>## [1] 0.2649877</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># The observed distribution of d's</span>
tab &lt;-<span class="st"> </span><span class="kw">prop.table</span>(<span class="kw">table</span>(x))
tab</code></pre></div>
<pre><code>## x
##   10   11   12   13   14   15   16 
## 0.01 0.03 0.15 0.55 0.20 0.05 0.01</code></pre>
<p>This can be plotted against the expected counts as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(1L:<span class="kw">length</span>(tab), <span class="kw"><a href="http://www.rdocumentation.org/packages/disclap/topics/ddisclap">ddisclap</a></span>(<span class="kw">as.integer</span>(<span class="kw">names</span>(tab)) -<span class="st"> </span>y_hat, p_hat), 
  <span class="dt">type =</span> <span class="st">"h"</span>, <span class="dt">col =</span> <span class="st">"#999999"</span>, <span class="dt">lend =</span> <span class="st">"butt"</span>, <span class="dt">lwd =</span> <span class="dv">50</span>, 
  <span class="dt">xlab =</span> <span class="st">"x, e.g. Y-STR allele"</span>, <span class="dt">ylab =</span> <span class="st">"Probability mass"</span>, <span class="dt">axes =</span> <span class="ot">FALSE</span>)
<span class="kw">axis</span>(<span class="dv">1</span>, <span class="dt">at =</span> 1L:<span class="kw">length</span>(tab), <span class="dt">labels =</span> <span class="kw">names</span>(tab))
<span class="kw">axis</span>(<span class="dv">2</span>)
<span class="kw">points</span>(1L:<span class="kw">length</span>(tab), tab, <span class="dt">type =</span> <span class="st">"h"</span>, <span class="dt">col =</span> <span class="st">"#000000"</span>, 
  <span class="dt">lend =</span> <span class="st">"butt"</span>, <span class="dt">lwd =</span> <span class="dv">25</span>)
<span class="kw">legend</span>(<span class="st">"topright"</span>, <span class="kw">c</span>(<span class="st">"Estimated distribution"</span>, <span class="st">"Observations"</span>), 
  <span class="dt">pch =</span> <span class="dv">15</span>, <span class="dt">col =</span> <span class="kw">c</span>(<span class="st">"#999999"</span>, <span class="st">"#000000"</span>))</code></pre></div>
<div class="figure">
<img src="introduction_files/figure-html/unnamed-chunk-7-1.png" alt="Observed frequencies of the $x$'s compared to a discrete Laplace distribution with parameters estimated from the sample." width="672"><p class="caption">
Observed frequencies of the <span class="math inline">\(x\)</span>’s compared to a discrete Laplace distribution with parameters estimated from the sample.
</p>
</div>
</div>
<div id="mixtures-of-discrete-laplace-distributions" class="section level1">
<h1 class="hasAnchor">
<a href="#mixtures-of-discrete-laplace-distributions" class="anchor"></a>Mixtures of discrete Laplace distributions</h1>
<p>In this section, the method is described. We do this first by a short formulation followed by a longer description with more details.</p>
<div id="short-model-formulation" class="section level2">
<h2 class="hasAnchor">
<a href="#short-model-formulation" class="anchor"></a>Short model formulation</h2>
Assume that there are <span class="math inline">\(c\)</span> subpopulations and that <span class="math inline">\(\tau_j = P(\text{From subpopulation } j)\)</span> is the a priori probability of a haplotype originated from the <span class="math inline">\(j\)</span>’th subpopulation for <span class="math inline">\(j = 1, 2, ..., c\)</span>. Then
<span class="math display">\[\begin{align}
    P(\text{Haplotype} = x \mid \text{From subpopulation } j)
\end{align}\]</span>
<p>is modelled by assuming independent discrete Laplace distribtions on loci as just described and by <span class="citation">(Andersen, Eriksen, and Morling 2013)</span>. The parameters of the model can be estimated using this package, <code>disclapmix</code>, as will be shown below.</p>
The haplotype frequency is obtained by summing the contribution from each subpopulation, such that
<span class="math display">\[\begin{align}  {#eq:vij}
    P(\text{Haplotype} = x) 
    = \sum_{j=1}^c \tau_j P(\text{Haplotype} = x \mid \text{From subpopulation } j) .
\end{align}\]</span>
</div>
<div id="longer-model-formulation" class="section level2">
<h2 class="hasAnchor">
<a href="#longer-model-formulation" class="anchor"></a>Longer model formulation</h2>
<p>Assume a very simple ‘haplotype’ with only one locus. Also assume a simple and isolated population. Then, it is reasonable to assume that there is a modal/central Y-STR allele, <span class="math inline">\(y\)</span>, and that all the alleles are distributed around this allele.</p>
<p>If we go back to Figure~@ref(fig:simple-example), this can be illustrated by <span class="math inline">\(y=13\)</span> as the central Y-STR allele and a distribution around <span class="math inline">\(y=13\)</span> with shorter and longer alleles.</p>
<p>To begin with, it might seem a bit overwhelming that Y-STR alleles should follow a simple probabiity distribution such as the discrete Laplace distribution. But surprisingly, it is actually a good approximation as demonstrated by <span class="citation">(Andersen, Eriksen, and Morling 2013)</span>.</p>
<p>We have haplotypes with several loci. When we assess multiple loci haplotypes, we assume that mutations happen independently across loci. Each locus has its own discrete Laplace distribution of allele probabilities, and the probability of a haplotype is the product of probabilities across loci. This gives a multivariate discrete Laplace distribution, where the marginals (that is, at each locus) are independent, discrete Laplace distributions.</p>
<p>Just as before, for a one locus haplotype, we can assume that there is a modal/central Y-STR profile with <span class="math inline">\(r\)</span> loci, <span class="math inline">\(y=(y_1, y_2, ..., y_r)\)</span>, and all the alleles are distributed around this profile. We also assume that the discrete Laplace distribution at each locus has its own parameter, where <span class="math inline">\(p_k\)</span> is the parameter at the <span class="math inline">\(k\)</span> locus. Normally, the central Y-STR profile, <span class="math inline">\(y\)</span>, would also be regarded as parameters.</p>
As before, let <span class="math inline">\(f(x; p, y)\)</span> be the probability mass function of a discrete Laplace distribution. We define an observation <span class="math inline">\(X = (X_{1}, X_{2}, ..., X_{r})\)</span> to be from a multivariate distribution of independent, discrete Laplace distributions when the probability of observing <span class="math inline">\(X=x\)</span> is
<span class="math display">\[\begin{align}  {#eq:disclap-independent}
  P(X = x) = \prod_{k=1}^r f \left ( x_{k} ; p_{k}, y_{k} \right ).
\end{align}\]</span>
<p>This corresponds to that the individual <span class="math inline">\(X\)</span> has mutated away from <span class="math inline">\(y\)</span> independently at each locus.</p>
Now, we have one more generalisation. A population may have several subpopulations, e.g. introduced by migration or by evolution. This means that we need to have a mixture of multivariate distributions with marginally independent, discrete Laplace distributions. Each component in the mixture represents a subpopulation. We define an observation <span class="math inline">\(X = (X_{1}, X_{2}, ..., X_{r})\)</span> to be from a mixture of multivariate, marginally independent, discrete Laplace distributions, when the probability of observing <span class="math inline">\(X=x\)</span> is
<span class="math display">\[\begin{align}  {#eq:disclap-mixtures}
  P(X = x) = \sum_{j=1}^c \tau_j \prod_{k=1}^r f \left ( x_{k} ; p_{jk}, y_{jk} \right ) ,
\end{align}\]</span>
<p>where <span class="math inline">\(\tau_j\)</span> is the a priori probability for originating from the <span class="math inline">\(j\)</span>’th subpopulation. Thus, the parameters of this mixture model are <span class="math inline">\(\{ y_j \}_{j=1}^c\)</span> with <span class="math inline">\(y_j = (y_{j1}, y_{j2}, ..., y_{jr})\)</span> as the central haplotype of the <span class="math inline">\(j\)</span> subpopulation, <span class="math inline">\(\{ \tau_j \}_{j=1}^c\)</span> and <span class="math inline">\(\{ p_{jk} \}_{\substack{j\in \{1, 2, ..., c\}\\k \in \{1, 2, ..., r\}}}\)</span> (the parameters for each discrete Laplace distribution).</p>
<p>We assume that <span class="math inline">\(p_{jk}\)</span> depends on locus and subpopulation, such that <span class="math inline">\(\log p_{jk} = \omega_j + \lambda_k\)</span>. This means that there is an additive effect of locus, <span class="math inline">\(\lambda_k\)</span>, and an additive effect of subpopulation, <span class="math inline">\(\omega_j\)</span>.</p>
<p>More theory on finite mixture distributions is given by <span class="citation">(Titterington, Smith, and Makov 1987)</span>.</p>
</div>
<div id="haplotype-frequency-prediction" class="section level2">
<h2 class="hasAnchor">
<a href="#haplotype-frequency-prediction" class="anchor"></a>Haplotype frequency prediction</h2>
<p>When we have estimated the parameters of a mixture of multivariate, marginally independent, discrete Laplace distributions (this will be shown in the next section), we can use these to estimate haplotype frequencies.</p>
Given estimates of subpopulation central haplotypes <span class="math inline">\(\{ \hat{y}_{j} \}_{j}\)</span>, dispersion parameters <span class="math inline">\(\{ \hat{p}_{jk} \}_{j,k}\)</span> and prior probabilities <span class="math inline">\(\{ \hat{\tau}_{j} \}_{j}\)</span>, the haplotype frequency of a haplotype <span class="math inline">\(x = (x_1, x_2, ..., x_r)\)</span> with <span class="math inline">\(x_k \in \mathbb{Z}\)</span> for <span class="math inline">\(k \in \{1, 2, ..., r\}\)</span> can be estimated as
<span class="math display">\[\begin{align}  {#eq:disclap-predict}
   \hat{p}(x) = \sum_{j=1}^c \hat{\tau}_j \prod_{k=1}^r f \left ( x_k ; \hat{p}_{jk}, \hat{y}_{jk} \right ).
\end{align}\]</span>
<p>Thus, we simply use the estimated parameters in Equation~ to obtain Equation~@ref(eq:disclap-predict).</p>
</div>
</div>
<div id="estimating-parameters" class="section level1">
<h1 class="hasAnchor">
<a href="#estimating-parameters" class="anchor"></a>Estimating parameters</h1>
<p>In this section we demonstrate how to estimate the parameters in a mixture of multivariate, independent, discrete Laplace distributions. This can for example be used to estimate Y-STR haplotype frequencies.</p>
<p>First, the <code>R</code> package <code>disclapmix</code> <span class="citation">(Andersen, Eriksen, and Morling 2013)</span> for analysing a mixture of multivariate, independent, discrete Laplace distributions must be loaded:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(disclapmix)</code></pre></div>
<p>If you do not have the <code>disclapmix</code> package installed, please visit <a href="http://cran.r-project.org/package=disclapmix" class="uri">http://cran.r-project.org/package=disclapmix</a>.</p>
<p>This package supplies the function <code>disclapmix</code> for estimating the parameters in a mixture of multivariate, marginally independent, discrete Laplace distributions with probability mass function given in Equation~. We will refer to this as ‘the discrete Laplace method’.</p>
<div id="data-from-marginally-independent-discrete-laplace-distributions" class="section level2">
<h2 class="hasAnchor">
<a href="#data-from-marginally-independent-discrete-laplace-distributions" class="anchor"></a>Data from marginally independent, discrete Laplace distributions</h2>
<p>Now, we revisit the example leading to Figure~@ref(fig:simple-example) and add two more loci with different dispersion and location parameters. We then analyse the randomly generated values from independent, discrete Laplace distributions with a probability mass function as given in Equation~.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1</span>)
n &lt;-<span class="st"> </span><span class="dv">100</span> <span class="co"># number of individuals</span>

<span class="co"># Locus 1</span>
p1 &lt;-<span class="st"> </span><span class="fl">0.3</span> <span class="co"># Dispersion parameter </span>
m1 &lt;-<span class="st"> </span>13L <span class="co"># Location parameter (L means integer)</span>
d1 &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/disclap/topics/ddisclap">rdisclap</a></span>(n, p1) +<span class="st"> </span>m1 <span class="co"># Generate a sampling using the rdisclap function</span>

<span class="co"># Locus 2</span>
p2 &lt;-<span class="st"> </span><span class="fl">0.4</span>
m2 &lt;-<span class="st"> </span>14L
d2 &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/disclap/topics/ddisclap">rdisclap</a></span>(n, p2) +<span class="st"> </span>m2

<span class="co"># Locus 3</span>
p3 &lt;-<span class="st"> </span><span class="fl">0.5</span>
m3 &lt;-<span class="st"> </span>15L
d3 &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/disclap/topics/ddisclap">rdisclap</a></span>(n, p3) +<span class="st"> </span>m3

db &lt;-<span class="st"> </span><span class="kw">cbind</span>(d1, d2, d3)
<span class="kw">head</span>(db)</code></pre></div>
<pre><code>##      d1 d2 d3
## [1,] 14 15 16
## [2,] 12 12 17
## [3,] 13 13 15
## [4,] 13 13 15
## [5,] 14 12 15
## [6,] 13 15 15</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit &lt;-<span class="st"> </span><span class="kw">disclapmix</span>(db, <span class="dt">clusters =</span> 1L) <span class="co"># L means integer type</span></code></pre></div>
<p>We can then look at the estimated location parameters, <span class="math inline">\(y=(y_1, y_2, y_3)\)</span>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit$y</code></pre></div>
<pre><code>##      d1 d2 d3
## [1,] 13 14 15</code></pre>
<p>And the estimated dispersion parameters, <span class="math inline">\((p_1, p_2, p_3)\)</span>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit$disclap_parameters</code></pre></div>
<pre><code>##                 d1        d2        d3
## cluster1 0.2649877 0.4369163 0.5167462</code></pre>
<p>As seen, the estimated dispersion location parameters are well estimated. The dispersion parameters are also quite close to the ones used to generate the data.</p>
</div>
<div id="data-from-a-fisher-wright-population" class="section level2">
<h2 class="hasAnchor">
<a href="#data-from-a-fisher-wright-population" class="anchor"></a>Data from a Fisher-Wright population</h2>
<p><span class="citation">(Andersen, Eriksen, and Morling 2013)</span> simulated populations following the Fisher-Wright model of evolution <span class="citation">(Fisher 1922; Fisher 1930; Fisher 1958; Wright 1931; Ewens 2004)</span> with assumptions of primarily neutral, single-step mutations of STRs <span class="citation">(Ohta and Kimura 1973)</span>. From these populations, data sets were sampled. Using the discrete Laplace method for estimating haplotype frequencies, the method worked rather well.</p>
<p>This is worth highlighting: Data was simulated under a completely different model than that used for inference afterwards. The data was simulated under a population model (Fisher-Wright model of evolution) with a certain mutation model (single-step mutation model). Inference was made assuming that the data was from a mixture of multivariate, marginally independent, discrete Laplace distributions.</p>
<p>One of the reasons that the discrete Laplace distribution predicts data from a Fisher-Wright model of evolution with a single-step mutation model is due to the fact that it approximates certain properties of this population and mutation model <span class="citation">(A. Caliebe et al. 2010)</span>. This is also explained by <span class="citation">(Andersen, Eriksen, and Morling 2013)</span>.</p>
<p>Now, let us try simulating a Fisher-Wright population and analyse it with the discrete Laplace method. To simulate the population, the <code>R</code> package <code>fwsim</code> <span class="citation">(Andersen and Eriksen 2015; Andersen and Eriksen 2012)</span> is loaded:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(fwsim)</code></pre></div>
<p>If you do not have the <code>fwsim</code> package installed, please visit <a href="http://cran.r-project.org/package=fwsim" class="uri">http://cran.r-project.org/package=fwsim</a>.</p>
<p>We then simulate a population consisting of Y-STR profiles:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1</span>)
generations &lt;-<span class="st"> </span>100L
pop_size &lt;-<span class="st"> </span>1e5L
loci &lt;-<span class="st"> </span>7L
mutation_rates &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="fl">0.001</span>, <span class="fl">0.01</span>, <span class="dt">length.out =</span> loci)
mutation_rates</code></pre></div>
<pre><code>## [1] 0.0010 0.0025 0.0040 0.0055 0.0070 0.0085 0.0100</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sim &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/fwsim/topics/fwsim">fwsim</a></span>(<span class="dt">G =</span> generations, <span class="dt">H0 =</span> <span class="kw">rep</span>(0L, loci), <span class="dt">N0 =</span> pop_size, 
  <span class="dt">mutmodel =</span> mutation_rates, <span class="dt">progress =</span> <span class="ot">FALSE</span>)
<span class="kw">summary</span>(sim)</code></pre></div>
<pre><code>## Fisher-Wright population with stochastic population size.
## Parameters:
##   G        = 100
##   progress = FALSE
##   trace    = FALSE
##   alpha    = 1, 1, 1, 1, 1, 1, ...
## 
## H0 and N0:
##   Locus1 Locus2 Locus3 Locus4 Locus5 Locus6 Locus7     N0
## 1      0      0      0      0      0      0      0 100000
## 
## SMM (stepwise mutation model) on 7 loci with parameters:
##      Locus1  Locus2 Locus3  Locus4 Locus5  Locus6 Locus7
## mu_d  5e-04 0.00125  0.002 0.00275 0.0035 0.00425  0.005
## mu_u  5e-04 0.00125  0.002 0.00275 0.0035 0.00425  0.005
## 
## Number of saved populations = 0
## 
## Final population (size = 101736):
##   Locus1 Locus2 Locus3 Locus4 Locus5 Locus6 Locus7 N
## 1      0     -1      1      0      3     -2      0 5
## 2      0      1      0      3     -1      0     -3 2
## 3      0      0      1      1      1      2      1 1
## 4      0     -1      1      1      0      1      1 3
## 5      0     -1     -1     -1      1      1      0 5
##  (4383 rows hidden.)
## 
## Expected population sizes = 1e+05, 1e+05, 1e+05, 1e+05, 1e+05, 1e+05, ...
## Actual population sizes   = 99801, 99979, 99613, 99476, 99781, 99764, ...</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pop &lt;-<span class="st"> </span>sim$population</code></pre></div>
<p>Note, that the mutation rates are different for each locus (ranging from  to ). The location parameter is set to 0 for all loci. This can be changed afterwards without loosing or adding any information. Below, we change it to be <span class="math inline">\(y = (14, 12, 28, 22, 10, 11, 13)\)</span>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">y &lt;-<span class="st"> </span><span class="kw">c</span>(14L, 12L, 28L, 22L, 10L, 11L, 13L)
for (i in 1L:loci) {
  pop[, i] &lt;-<span class="st"> </span>pop[, i] +<span class="st"> </span>y[i]
}
<span class="kw">head</span>(pop)</code></pre></div>
<pre><code>##   Locus1 Locus2 Locus3 Locus4 Locus5 Locus6 Locus7 N
## 1     14     11     29     22     13      9     13 5
## 2     14     13     28     25      9     11     10 2
## 3     14     12     29     23     11     13     14 1
## 4     14     11     29     23     10     12     14 3
## 5     14     11     27     21     11     12     13 5
## 6     14     12     29     22     11     13     14 3</code></pre>
<p>Then, <span class="math inline">\(y\)</span> is the most frequent 10 locus Y-STR haplotype in Denmark according to <a href="http://www.yhrd.org" class="uri">http://www.yhrd.org</a> (on March 26, 2013) restricted to the 7 loci minimal haplotype.</p>
<p>The column <code>N</code> is the number of individuals in the population with that Y-STR haplotype. Summing column <code>N</code> reveals that there is not exactly <code>pop\_size</code> individuals due to that the population size is stochastic (refer to <span class="citation">(Andersen and Eriksen 2012)</span> for the details).</p>
<p>We can then calculate the population frequency for each haplotype:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pop$PopFreq &lt;-<span class="st"> </span>pop$N /<span class="st"> </span><span class="kw">sum</span>(pop$N)</code></pre></div>
<p>Let us draw a data set where each haplotype is drawn relatively to its population frequency:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1</span>)
n &lt;-<span class="st"> </span><span class="dv">500</span> <span class="co"># Data set size</span>
types &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="dt">x =</span> 1L:<span class="kw">nrow</span>(pop), <span class="dt">size =</span> n, <span class="dt">replace =</span> <span class="ot">TRUE</span>, <span class="dt">prob =</span> pop$N)
types_table &lt;-<span class="st"> </span><span class="kw">table</span>(types)

alpha &lt;-<span class="st"> </span><span class="kw">sum</span>(types_table ==<span class="st"> </span><span class="dv">1</span>) 
alpha /<span class="st"> </span>n <span class="co"># Singleton proportion</span></code></pre></div>
<pre><code>## [1] 0.462</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dataset &lt;-<span class="st"> </span>pop[<span class="kw">as.integer</span>(<span class="kw">names</span>(types_table)), ]
dataset$Ndb &lt;-<span class="st"> </span>types_table
<span class="kw">head</span>(dataset)</code></pre></div>
<pre><code>##     Locus1 Locus2 Locus3 Locus4 Locus5 Locus6 Locus7  N      PopFreq Ndb
## 9       14     12     29     22     10     12     15 15 0.0001474404   2
## 38      14     12     27     22      9     12     13 94 0.0009239601   1
## 80      14     11     27     22     10     11     12 40 0.0003931745   1
## 159     14     12     28     23      9     10     13 73 0.0007175434   2
## 168     14     12     28     22      9     13     14 60 0.0005897617   1
## 181     13     12     28     22      9     11     13 67 0.0006585673   1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">db &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(pop[types, 1L:loci])
<span class="kw">head</span>(db)</code></pre></div>
<pre><code>##      Locus1 Locus2 Locus3 Locus4 Locus5 Locus6 Locus7
## 1166     14     11     28     22     10      8     14
## 3186     14     12     28     21      9     12     12
## 1763     14     12     28     22     10     10     12
## 488      14     12     28     22      9     11     13
## 3794     14     12     28     22     11     12     13
## 3943     14     12     28     22     10     12     13</code></pre>
<p>Then, analyse it:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit &lt;-<span class="st"> </span><span class="kw">disclapmix</span>(db, <span class="dt">clusters =</span> 1L)

<span class="co"># Estimated location parameters</span>
fit$y </code></pre></div>
<pre><code>##      Locus1 Locus2 Locus3 Locus4 Locus5 Locus6 Locus7
## [1,]     14     12     28     22     10     11     13</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Estimated dispersion parameters</span>
fit$disclap_parameters</code></pre></div>
<pre><code>##              Locus1    Locus2   Locus3    Locus4    Locus5    Locus6
## cluster1 0.02797808 0.1164012 0.151445 0.2111471 0.2477904 0.2895135
##             Locus7
## cluster1 0.3010092</code></pre>
Let us compare the mutation rates with the dispersion parameters in the discrete Laplace distributions:

</div>
<div id="data-from-a-mixture-of-two-fisher-wright-populations" class="section level2">
<h2 class="hasAnchor">
<a href="#data-from-a-mixture-of-two-fisher-wright-populations" class="anchor"></a>Data from a mixture of two Fisher-Wright populations</h2>
<p>Here, we show how to analyse a dataset from a mixture of two populations. First, we simulate two populations (note the different mutation rates and location parameters, where the location parameters again are changed afterwards without loosing or adding any information):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1</span>)

<span class="co"># Common parameters</span>
generations &lt;-<span class="st"> </span>100L
pop_size &lt;-<span class="st"> </span>1e5L
loci &lt;-<span class="st"> </span>7L

mu1 &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="fl">0.001</span>, <span class="fl">0.005</span>, <span class="dt">length.out =</span> loci)
sim1 &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/fwsim/topics/fwsim">fwsim</a></span>(<span class="dt">G =</span> generations, <span class="dt">H0 =</span> y, <span class="dt">N0 =</span> pop_size, <span class="dt">mutmodel =</span> mu1, <span class="dt">progress =</span> <span class="ot">FALSE</span>)
pop1 &lt;-<span class="st"> </span>sim1$population

mu2 &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="fl">0.005</span>, <span class="fl">0.01</span>, <span class="dt">length.out =</span> loci)
sim2 &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/fwsim/topics/fwsim">fwsim</a></span>(<span class="dt">G =</span> generations, <span class="dt">H0 =</span> <span class="kw">c</span>(14L, 13L, 29L, 23L, 11L, 13L, 13L), <span class="dt">N0 =</span> pop_size, <span class="dt">mutmodel =</span> mu2, <span class="dt">progress =</span> <span class="ot">FALSE</span>)
pop2 &lt;-<span class="st"> </span>sim2$population</code></pre></div>
<p>Here, just as <code>y</code> are the alleles from most frequent haplotype, then <span class="math inline">\((14, 13, 29, 23, 11, 13, 13)\)</span> are the alleles from the second most frequent haplotype.</p>
<p>Then we sample a data set with an expected proportion of 20% from the first population and 80% from the second population:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1</span>)
n &lt;-<span class="st"> </span>500L <span class="co"># Data set size</span>

n1 &lt;-<span class="st"> </span><span class="kw">rbinom</span>(<span class="dv">1</span>, n, <span class="fl">0.2</span>)
<span class="kw">c</span>(n1, n1 /<span class="st"> </span>n)</code></pre></div>
<pre><code>## [1] 102.000   0.204</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">n2 &lt;-<span class="st"> </span>n -<span class="st"> </span>n1
<span class="kw">c</span>(n2, n2 /<span class="st"> </span>n)</code></pre></div>
<pre><code>## [1] 398.000   0.796</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">types1 &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="dt">x =</span> 1L:<span class="kw">nrow</span>(pop1), <span class="dt">size =</span> n1, <span class="dt">replace =</span> <span class="ot">TRUE</span>, <span class="dt">prob =</span> pop1$N)
db1 &lt;-<span class="st"> </span>pop1[types1, 1L:loci]

types2 &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="dt">x =</span> 1L:<span class="kw">nrow</span>(pop2), <span class="dt">size =</span> n2, <span class="dt">replace =</span> <span class="ot">TRUE</span>, <span class="dt">prob =</span> pop2$N)
db2 &lt;-<span class="st"> </span>pop2[types2, 1L:loci]

db &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(<span class="kw">rbind</span>(db1, db2))

<span class="co"># Singleton proportion</span>
<span class="kw">sum</span>(<span class="kw">table</span>(<span class="kw">apply</span>(db, <span class="dv">1</span>, paste, <span class="dt">collapse =</span> <span class="st">";"</span>)) ==<span class="st"> </span>1L) /<span class="st"> </span>n</code></pre></div>
<pre><code>## [1] 0.692</code></pre>
<p>Now, we analyse the data set trying 1 to 5 subpopulations. Afterwards, we analyse the optimal number of subpopulations using the BIC (Bayesian Information Criteria) by <span class="citation">(Schwarz 1978)</span>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fits &lt;-<span class="st"> </span><span class="kw">lapply</span>(1L:5L, function(clusters) <span class="kw">disclapmix</span>(db, <span class="dt">clusters =</span> clusters, <span class="dt">iterations =</span> 100L))</code></pre></div>
<p>The BIC values are:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">BIC &lt;-<span class="st"> </span><span class="kw">sapply</span>(fits, function(fit) fit$BIC_marginal)</code></pre></div>
<p>The estimated parameters for this optimal number of subpopulations can be made available in <code>best\_fit</code> as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">best_fit &lt;-<span class="st"> </span>fits[[<span class="kw">which.min</span>(BIC)]]
best_fit</code></pre></div>
<pre><code>## disclapmixfit from 500 observations on 7 loci with 2 clusters.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Estimated a priori probability of originating from each subpopulation</span>
best_fit$tau </code></pre></div>
<pre><code>## [1] 0.2190092 0.7809908</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Estimated location parameters</span>
best_fit$y </code></pre></div>
<pre><code>##      Locus1 Locus2 Locus3 Locus4 Locus5 Locus6 Locus7
## [1,]     14     12     28     22     10     11     13
## [2,]     14     13     29     23     11     13     13</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Estimated dispersion parameters for each subpopulation</span>
best_fit$disclap_parameters</code></pre></div>
<pre><code>##              Locus1    Locus2    Locus3    Locus4    Locus5    Locus6
## cluster1 0.09007892 0.1091864 0.1165970 0.1380979 0.1531467 0.1568696
## cluster2 0.16811193 0.2037718 0.2176019 0.2577285 0.2858136 0.2927617
##             Locus7
## cluster1 0.1806647
## cluster2 0.3371698</code></pre>
<p>The estimated location parameters are the same as those used for generating the data. Also, the values of <span class="math inline">\(\tau_j\)</span>, the a priori probability of originating from the <span class="math inline">\(j\)</span> subpopulation, are consistent with the mixture proportions of  and .</p>
<p>We can also calculate the predicted population frequencies (using the mixture proportions  and ):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pop1$PopFreq &lt;-<span class="st"> </span>pop1$N /<span class="st"> </span><span class="kw">sum</span>(pop1$N)
pop2$PopFreq &lt;-<span class="st"> </span>pop2$N /<span class="st"> </span><span class="kw">sum</span>(pop2$N)

types1.table &lt;-<span class="st"> </span><span class="kw">table</span>(types1)
types2.table &lt;-<span class="st"> </span><span class="kw">table</span>(types2)

dataset1 &lt;-<span class="st"> </span>pop1[<span class="kw">as.integer</span>(<span class="kw">names</span>(types1.table)), ]
dataset1$Ndb &lt;-<span class="st"> </span>types1.table
<span class="kw">sum</span>(dataset1$Ndb)</code></pre></div>
<pre><code>## [1] 102</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dataset2 &lt;-<span class="st"> </span>pop2[<span class="kw">as.integer</span>(<span class="kw">names</span>(types2.table)), ]
dataset2$Ndb &lt;-<span class="st"> </span>types2.table
<span class="kw">sum</span>(dataset2$Ndb)</code></pre></div>
<pre><code>## [1] 398</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dataset &lt;-<span class="st"> </span><span class="kw">merge</span>(<span class="dt">x =</span> dataset1, <span class="dt">y =</span> dataset2, <span class="dt">by =</span> <span class="kw">colnames</span>(db), <span class="dt">all =</span> <span class="ot">TRUE</span>)
dataset[<span class="kw">is.na</span>(dataset)] &lt;-<span class="st"> </span><span class="dv">0</span>

dataset$MixPopFreq &lt;-<span class="st"> </span>(n1/n) *<span class="st"> </span>dataset$PopFreq.x +<span class="st"> </span>(n2/n) *<span class="st"> </span>dataset$PopFreq.y

dataset$Type &lt;-<span class="st"> "Only from pop1"</span>
dataset$Type[dataset$Ndb.y &gt;<span class="st"> </span><span class="dv">0</span>] &lt;-<span class="st"> "Only from pop2"</span>
dataset$Type[dataset$Ndb.x &gt;<span class="st"> </span><span class="dv">0</span> &amp;<span class="st"> </span>dataset$Ndb.y &gt;<span class="st"> </span><span class="dv">0</span>] &lt;-<span class="st"> "Occurred in both"</span>
dataset$Type &lt;-<span class="st"> </span><span class="kw">factor</span>(dataset$Type)</code></pre></div>
We can now compare the predicted frequencies with the population frequency:

</div>
</div>
<div id="advanced-topics" class="section level1">
<h1 class="hasAnchor">
<a href="#advanced-topics" class="anchor"></a>Advanced topics</h1>
<p>In this section, more advanced topics are demonstrated.</p>
<div id="database-spectrum" class="section level2">
<h2 class="hasAnchor">
<a href="#database-spectrum" class="anchor"></a>Database spectrum</h2>
<p>The spectrum of a database is the number of haplotypes occuring once, the number occuring twice etc. Here, we investigate how subsampling can be used to assess the variability of the database spectrum. We use <code>best\_fit</code> from the mixture of two Fisher-Wright populations:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">best_fit</code></pre></div>
<pre><code>## disclapmixfit from 500 observations on 7 loci with 2 clusters.</code></pre>
<p>Let us simulate 100 more databases from the fitted discrete Laplace model and see the number of singletons (other aspects of the database spectrum are interesting, too, but here we focus on the number of singletons):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1</span>)
sim_dbs &lt;-<span class="st"> </span><span class="kw">lapply</span>(1L:100L, function(i) {
  <span class="kw">simulate</span>(best_fit, <span class="dt">nsim =</span> <span class="kw">nrow</span>(db))
})
<span class="kw">head</span>(sim_dbs[[<span class="dv">1</span>]])</code></pre></div>
<pre><code>##      [,1] [,2] [,3] [,4] [,5] [,6] [,7]
## [1,]   14   13   29   23    9   14   14
## [2,]   14   13   29   23   10   10   13
## [3,]   14   14   29   24   12   13   13
## [4,]   12   13   29   23    9   14   13
## [5,]   14   14   28   23   11   13   12
## [6,]   14   13   29   22   11   13   12</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">num_singletons &lt;-<span class="st"> </span>function(d) {
  <span class="kw">sum</span>(<span class="kw">table</span>(<span class="kw">apply</span>(d, <span class="dv">1</span>, paste, <span class="dt">collapse =</span> <span class="st">";"</span>)) ==<span class="st"> </span>1L)
}

sim_dbs_singletons &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>(sim_dbs, function(sim_db) { 
  <span class="kw">num_singletons</span>(sim_db)
}))

<span class="kw">summary</span>(sim_dbs_singletons)</code></pre></div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##   325.0   342.8   351.0   351.5   360.0   379.0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">num_singletons</span>(db)</code></pre></div>
<pre><code>## [1] 346</code></pre>
</div>
<div id="simulating-more-advanced-populations" class="section level2">
<h2 class="hasAnchor">
<a href="#simulating-more-advanced-populations" class="anchor"></a>Simulating more advanced populations</h2>
<p>We can simulate more advanced populations by 1) using a more sophisticated mutation model and 2) having more than one initial haplotype. Here, we use the logistic mutation model introduced by <span class="citation">(A. Jochens et al. 2011)</span>. First, let us inspect the stationary distribution (how the allele distribution would look like in the limit of an infinite number of generations):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mutpars.locus1 &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="fl">0.149</span>,   <span class="fl">2.08</span>,    <span class="fl">18.3</span>,   <span class="fl">0.149</span>,   <span class="fl">0.374</span>,   <span class="fl">27.4</span>) <span class="co"># DYS19</span>
mutpars.locus2 &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="fl">0.500</span>,   <span class="fl">1.18</span>,    <span class="fl">18.0</span>,   <span class="fl">0.500</span>,   <span class="fl">0.0183</span>,  <span class="dv">349</span>)  <span class="co"># DYS389I</span>
mutpars.locus3 &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="fl">0.0163</span>,  <span class="fl">17.7</span>,    <span class="fl">11.1</span>,   <span class="fl">0.0163</span>,  <span class="fl">0.592</span>,   <span class="fl">14.1</span>)  <span class="co"># DYS391</span>
mutpars &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">c</span>(mutpars.locus1, mutpars.locus2, mutpars.locus3), <span class="dt">ncol =</span> <span class="dv">3</span>)
<span class="kw">colnames</span>(mutpars) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"DYS19"</span>, <span class="st">"DYS389I"</span>, <span class="st">"DYS391"</span>)
mutmodel &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/fwsim/topics/init_mutmodel">init_mutmodel</a></span>(<span class="dt">modeltype =</span> 2L, <span class="dt">mutpars =</span> mutpars)

<span class="kw"><a href="http://www.rdocumentation.org/packages/fwsim/topics/mutmodel">mutmodel_not_mut</a></span>(mutmodel, <span class="dt">locus =</span> 1L, <span class="dt">alleles =</span> 5L:20L)</code></pre></div>
<pre><code>##         5         6         7         8         9        10        11 
## 0.9999657 0.9999502 0.9999276 0.9998948 0.9998472 0.9997780 0.9996775 
##        12        13        14        15        16        17        18 
## 0.9995315 0.9993180 0.9989947 0.9984159 0.9966968 0.9876656 0.9437162 
##        19        20 
## 0.8730008 0.8464115</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/fwsim/topics/mutmodel">mutmodel_dw_mut</a></span>(mutmodel, <span class="dt">locus =</span> 1L, <span class="dt">alleles =</span> 5L:20L)</code></pre></div>
<pre><code>##            5            6            7            8            9 
## 1.441663e-13 1.153975e-12 9.236954e-12 7.393691e-11 5.918257e-10 
##           10           11           12           13           14 
## 4.737250e-09 3.791916e-08 3.035222e-07 2.429500e-06 1.944463e-05 
##           15           16           17           18           19 
## 1.555018e-04 1.235677e-03 9.347923e-03 5.198197e-02 1.208271e-01 
##           20 
## 1.447825e-01</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/fwsim/topics/mutmodel">mutmodel_up_mut</a></span>(mutmodel, <span class="dt">locus =</span> 1L, <span class="dt">alleles =</span> 5L:20L)</code></pre></div>
<pre><code>##            5            6            7            8            9 
## 3.425634e-05 4.978767e-05 7.235726e-05 1.051508e-04 1.527917e-04 
##           10           11           12           13           14 
## 2.219852e-04 3.224458e-04 4.682274e-04 6.796174e-04 9.858098e-04 
##           15           16           17           18           19 
## 1.428624e-03 2.067568e-03 2.986491e-03 4.301870e-03 6.172108e-03 
##           20 
## 8.805949e-03</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">statdists &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/fwsim/topics/mutmodel">approx_stationary_dist</a></span>(mutmodel, <span class="dt">alleles =</span> 5L:20L)
bp &lt;-<span class="st"> </span><span class="kw">barplot</span>(statdists, <span class="dt">beside =</span> <span class="ot">TRUE</span>, <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">1.1</span>*<span class="kw">max</span>(statdists)))
<span class="kw">axis</span>(<span class="dv">3</span>, bp, <span class="kw">rep</span>(<span class="kw">rownames</span>(statdists), <span class="kw">ncol</span>(mutmodel$mutpars)), <span class="dt">cex.axis =</span> <span class="dv">1</span>, <span class="dt">las =</span> <span class="dv">3</span>)</code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-29-1.png" width="\textwidth"> Next, let us simulate a population with two starting haplotypes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rare_haplotype &lt;-<span class="st"> </span><span class="kw">c</span>(12L, 9L, 9L) <span class="co"># 0/126931 at yhrd.org r46</span>
common_haplotype &lt;-<span class="st"> </span><span class="kw">c</span>(14L, 12L, 10L) <span class="co"># 8541/126931 at yhrd.org r46</span>
H0 &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">c</span>(rare_haplotype, common_haplotype), <span class="dv">2</span>, <span class="dv">3</span>, <span class="dt">byrow =</span> <span class="ot">TRUE</span>)
<span class="kw">set.seed</span>(<span class="dv">1</span>)
sim &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/fwsim/topics/fwsim">fwsim</a></span>(<span class="dt">G =</span> 5000L, <span class="dt">H0 =</span> H0, <span class="dt">N0 =</span> <span class="kw">c</span>(1e6L, 1e6L), 
  <span class="dt">mutmodel =</span> mutmodel, <span class="dt">progress =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p>Let us inspect the simulation and the 10 most common haplotypes in the final population</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sim$population[<span class="kw">order</span>(sim$population$N, <span class="dt">decreasing =</span> <span class="ot">TRUE</span>)[1L:10L], ]</code></pre></div>
<pre><code>##     DYS19 DYS389I DYS391      N
## 596    16      12     10 157909
## 598    15      12     10 157669
## 589    16      13     10 116830
## 594    15      13     10 111012
## 430    15      12     11  88089
## 432    16      12     11  87543
## 462    16      11     10  69504
## 460    15      11     10  69212
## 506    15      13     11  64630
## 592    16      13     11  63882</code></pre>
<p>Let us compare the mutation probabilities for the initial haplotypes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="dv">1</span>-<span class="kw">sapply</span>(1L:<span class="kw">ncol</span>(mutmodel$mutpars), function(locus) {
  <span class="kw"><a href="http://www.rdocumentation.org/packages/fwsim/topics/mutmodel">mutmodel_not_mut</a></span>(mutmodel, <span class="dt">locus =</span> locus, <span class="dt">alleles =</span> rare_haplotype[locus])
})</code></pre></div>
<pre><code>##           12            9            9 
## 0.0004685310 0.0010028796 0.0007590255</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="dv">1</span>-<span class="kw">sapply</span>(1L:<span class="kw">ncol</span>(mutmodel$mutpars), function(locus) {
  <span class="kw"><a href="http://www.rdocumentation.org/packages/fwsim/topics/mutmodel">mutmodel_not_mut</a></span>(mutmodel, <span class="dt">locus =</span> locus, <span class="dt">alleles =</span> common_haplotype[locus])
})</code></pre></div>
<pre><code>##          14          12          10 
## 0.001005254 0.001466993 0.001322288</code></pre>
<p>As seen, the rare haplotype is almost not mutating. The rationale is that the alleles should mutate up before reaching as low as the alleles of the rare haplotype:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sapply</span>(1L:<span class="kw">ncol</span>(mutmodel$mutpars), function(locus) <span class="kw"><a href="http://www.rdocumentation.org/packages/fwsim/topics/mutmodel">mutmodel_dw_mut</a></span>(mutmodel, <span class="dt">locus =</span> locus, <span class="dt">alleles =</span> rare_haplotype[locus]))</code></pre></div>
<pre><code>##           12            9            9 
## 3.035222e-07 1.221102e-05 1.173442e-18</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sapply</span>(1L:<span class="kw">ncol</span>(mutmodel$mutpars), function(locus) <span class="kw"><a href="http://www.rdocumentation.org/packages/fwsim/topics/mutmodel">mutmodel_up_mut</a></span>(mutmodel, <span class="dt">locus =</span> locus, <span class="dt">alleles =</span> rare_haplotype[locus]))</code></pre></div>
<pre><code>##           12            9            9 
## 0.0004682274 0.0009906686 0.0007590255</code></pre>
<p>Let us check how many close to the rare haplotype that are left:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dists &lt;-<span class="st"> </span><span class="kw">apply</span>(sim$population[, -<span class="kw">ncol</span>(sim$population)], <span class="dv">1</span>, function(h) {
  <span class="kw">sum</span>(<span class="kw">abs</span>(h -<span class="st"> </span>rare_haplotype))
})
dists_order &lt;-<span class="st"> </span><span class="kw">order</span>(dists, <span class="dt">decreasing =</span> <span class="ot">FALSE</span>)[1L:10L]
<span class="kw">data.frame</span>(sim$population[dists_order, ], 
  <span class="dt">DistanceToRare =</span> dists[dists_order])</code></pre></div>
<pre><code>##     DYS19 DYS389I DYS391    N DistanceToRare
## 57     12      10      9 1350              1
## 86     12       9     10  688              1
## 7      13       9     10  345              2
## 34     12       8     10  163              2
## 56     12      11      9  703              2
## 58     12       9     11  162              2
## 62     12      10     10 4885              2
## 128    13      10      9  754              2
## 3      14      10      9  419              3
## 9      13      10     10 6009              3</code></pre>
</div>
</div>
<div id="concluding-remarks" class="section level1">
<h1 class="hasAnchor">
<a href="#concluding-remarks" class="anchor"></a>Concluding remarks</h1>
<p>We have shown how to analyse Y-STR population data using the discrete Laplace method described by <span class="citation">(Andersen, Eriksen, and Morling 2013)</span>. This was done using the freely available and open source <code>R</code> packages <code>disclap</code>, <code>fwsim</code> and <code>disclapmix</code> that are supported on Linux, MacOS and MS Windows.</p>
<p>One key point made is worth repeating: Data simulated under a population model (e.g. the Fisher-Wright model of evolution) with a certain mutation model (e.g. the single-step mutation model) can be successfully analysed using the discrete Laplace method making inference assuming that the data is from a mixture of multivariate, independent, discrete Laplace distributions.</p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-fwsimArxiv">
<p>Andersen, Mikkel Meyer, and Poul Svante Eriksen. 2012. “Efficient Forward Simulation of Fisher-Wright Populations with Stochastic Population Size and Neutral Single Step Mutations in Haplotypes.” <em>Preprint, ArXiv:1210.1773</em>.</p>
</div>
<div id="ref-disclap">
<p>———. 2014. <em>disclap: Discrete Laplace Exponential Family</em>. <a href="http://CRAN.R-project.org/package=disclap" class="uri">http://CRAN.R-project.org/package=disclap</a>.</p>
</div>
<div id="ref-fwsim">
<p>———. 2015. <em>Fwsim: Fisher-Wright Population Simulation</em>. <a href="http://CRAN.R-project.org/package=fwsim" class="uri">http://CRAN.R-project.org/package=fwsim</a>.</p>
</div>
<div id="ref-AndersenDisclap2013">
<p>Andersen, Mikkel Meyer, Poul Svante Eriksen, and Niels Morling. 2013. “The discrete Laplace exponential family and estimation of Y-STR haplotype frequencies.” <em>Journal of Theoretical Biology</em> 329: 39–51.</p>
</div>
<div id="ref-Caliebe2010">
<p>Caliebe, Amke, Arne Jochens, Michael Krawczak, and Uwe Rösler. 2010. “A Markov Chain Description of the Stepwise Mutation Model: Local and Global Behaviour of the Allele Process.” <em>Journal of Theoretical Biology</em> 266 (2): 336–42.</p>
</div>
<div id="ref-Ewens2004">
<p>Ewens, Warren J. 2004. <em>Mathematical Population Genetics</em>. Springer-Verlag.</p>
</div>
<div id="ref-Fisher1922">
<p>Fisher, R. A. 1922. “On the Dominance Ratio.” <em>Proc. Roy. Soc. Edin.</em> 42: 321–41.</p>
</div>
<div id="ref-Fisher1930">
<p>———. 1930. <em>The Genetical Theory of Natural Selection</em>. Oxford: Clarendon Press.</p>
</div>
<div id="ref-Fisher1958">
<p>———. 1958. <em>The Genetical Theory of Natural Selection</em>. 2nd revised. New York: Dover.</p>
</div>
<div id="ref-Jochens2011">
<p>Jochens, A., A. Caliebe, U. Rösler, and M. Krawczak. 2011. “Empirical Evaluation Reveals Best Fit of a Logistic Mutation Model for Human Y-chromosomal Microsatellites.” <em>Genetics</em>.</p>
</div>
<div id="ref-OhtaKimura1973">
<p>Ohta, T., and M. Kimura. 1973. “A Model of Mutation Appropriate to Estimate the Number of Electrophoretically Detectable Alleles in a Finite Population.” <em>Genet. Res.</em> 22: 201–4.</p>
</div>
<div id="ref-BIC">
<p>Schwarz, Gideon. 1978. “Estimating the Dimension of a Model.” <em>Annals of Statistics</em> 6 (2): 461–64.</p>
</div>
<div id="ref-Titterington1987">
<p>Titterington, D. M., A. F. M. Smith, and U. E. Makov. 1987. <em>Statistical Analysis of Finite Mixture Distributions</em>. Wiley.</p>
</div>
<div id="ref-Wright1931">
<p>Wright, S. 1931. “Evolution in Mendelian populations.” <em>Genetics</em> 16: 97–159.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#sec:disclap">The discrete Laplace distribution</a></li>
      <li>
<a href="#mixtures-of-discrete-laplace-distributions">Mixtures of discrete Laplace distributions</a><ul class="nav nav-pills nav-stacked">
<li><a href="#short-model-formulation">Short model formulation</a></li>
      <li><a href="#longer-model-formulation">Longer model formulation</a></li>
      <li><a href="#haplotype-frequency-prediction">Haplotype frequency prediction</a></li>
      </ul>
</li>
      <li>
<a href="#estimating-parameters">Estimating parameters</a><ul class="nav nav-pills nav-stacked">
<li><a href="#data-from-marginally-independent-discrete-laplace-distributions">Data from marginally independent, discrete Laplace distributions</a></li>
      <li><a href="#data-from-a-fisher-wright-population">Data from a Fisher-Wright population</a></li>
      <li><a href="#data-from-a-mixture-of-two-fisher-wright-populations">Data from a mixture of two Fisher-Wright populations</a></li>
      </ul>
</li>
      <li>
<a href="#advanced-topics">Advanced topics</a><ul class="nav nav-pills nav-stacked">
<li><a href="#database-spectrum">Database spectrum</a></li>
      <li><a href="#simulating-more-advanced-populations">Simulating more advanced populations</a></li>
      </ul>
</li>
      <li><a href="#concluding-remarks">Concluding remarks</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Mikkel Meyer Andersen, Poul Svante Eriksen.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
